common/flatpak-installation.c

```diff
@@ -1092,7 +1098,93 @@ flatpak_installation_list_installed_refs_for_update (FlatpakInstallation *self,
+          const gchar *runtime = NULL;
+          g_autofree gchar *full_runtime_ref = NULL;
+
+          installed_related_refs = flatpak_installation_list_installed_related_refs_sync (self,
+                                                                                          remote_name,
+                                                                                          full_ref,
+                                                                                          cancellable,
+                                                                                          &local_error);
+          if (local_error != NULL)
+            {
+              g_warning ("Unable to get installed related refs for %s: %s", full_ref, local_error->message);
+              g_clear_error (&local_error);
+              goto check_for_runtime;
+            }
+
+          remote_related_refs = flatpak_installation_list_remote_related_refs_sync (self,
+                                                                                    remote_name,
+                                                                                    full_ref,
+                                                                                    cancellable,
+                                                                                    &local_error);
+          if (local_error != NULL)
+            {
+              g_warning ("Unable to get remote related refs for %s: %s", full_ref, local_error->message);
+              g_clear_error (&local_error);
+              goto check_for_runtime;
+            }
+
+          if (remote_related_refs == NULL)
+            goto check_for_runtime;
+
+          for (j = 0; j < remote_related_refs->len; j++)
+            {
+              FlatpakRelatedRef *remote_related_ref = g_ptr_array_index (remote_related_refs, j);
+              gboolean should_download = flatpak_related_ref_should_download (remote_related_ref);
+              gboolean related_ref_found = FALSE;
+
+              if (should_download && installed_related_refs != NULL)
+                {
+                  for (k = 0; k < installed_related_refs->len; k++)
+                    {
+                      FlatpakRelatedRef *installed_related_ref = g_ptr_array_index (installed_related_refs, k);
+                      g_autofree gchar *full_ref1 = flatpak_ref_format_ref (FLATPAK_REF (installed_related_ref));
+                      g_autofree gchar *full_ref2 = flatpak_ref_format_ref (FLATPAK_REF (remote_related_ref));
+
+                      if (g_strcmp0 (full_ref1, full_ref2) == 0)
+                        {
+                          related_ref_found = TRUE;
+                          break;
+                        }
+                    }
+
+                  if (should_download && !related_ref_found)
+                    g_ptr_array_add (updates, g_object_ref (installed_ref));
+                }
+            }
+
+        check_for_runtime:
```

Comment:
I think that introducing a goto inside a loop (which itself contains a doubly-nested loop) in a 300-line function makes the control flow very hard to follow. I suggest moving this new logic into a function that could early-return in case of error, removing the need for the goto.